src/function/mod.rs

```diff
@@ -0,0 +1,78 @@
+use crate::{
+    error::{FreightError, OrReturn},
+    execution_engine::{evaluate, ExecutionEngine},
+    expression::Expression,
+    TypeSystem,
+};
+use std::fmt::Debug;
+
+mod arg_count;
+mod function_ref;
+mod function_type;
+mod function_writer;
+
+pub use arg_count::*;
+pub use function_ref::*;
+pub use function_type::*;
+pub use function_writer::*;
+
+#[derive(Debug)]
+pub struct Function<TS: TypeSystem> {
+    pub(crate) expressions: Vec<Expression<TS>>,
+
+    pub(crate) _stack_size: usize,
+    pub(crate) arg_count: ArgCount,
+    pub(crate) return_target: usize,
+}
+
+impl<TS: TypeSystem> Function<TS> {
+    pub fn call(
+        &self,
+        engine: &mut ExecutionEngine<TS>,
+        args: &mut [TS::Value],
+        captured: &[TS::Value],
+        // real:usize
+    ) -> Result<TS::Value, FreightError> {
+        match self.arg_count.max() {
+            Some(max) if max > args.len() => {
+                return Err(FreightError::IncorrectArgumentCount {
+                    expected_min: self.arg_count.min(),
+                    expected_max: self.arg_count.max(),
+                    actual: args.len(),
+                });
+            }
+            _ => (),
+        }
+        if self.expressions.is_empty() {
+            return Ok(Default::default());
+        }
+
+        #[cfg(feature = "variadic_functions")]
+        let mut args = match self.arg_count {
```

Comment:
What is going on here? Why are you returning a Result from a match statement only to immediately unwrap it with another match statement? Is there a way to represent this logic with a single match?

Reply:
I made the change to this I mentioned so it doesn't need the monad to hold state